src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java

```diff
@@ -0,0 +1,472 @@
+
+            //no work on those
+            final PCollection<Read> not_primary = readsPartitioned.get(ReadsPartition.NOT_PRIMARY.ordinal());
+
+            return PCollectionList.of(fragmentsTransformed).and(pairsTransformed).and(not_primary).apply(Flatten.<Read>pCollections());
+        }
+
+        private static PCollectionList<Read> partitionReadsByPrimaryNonPrimaryAlignment(PCollection<Read> preads) {
+            return preads.apply(Partition.of(2, new PartitionFn<Read>() {
+                private static final long serialVersionUID = 1l;
+
+                @Override
+                    public int partitionFor(final Read read, final int n) {
+                        return isPrimaryAlignment(read) ? ReadsPartition.PRIMARY.ordinal() : ReadsPartition.NOT_PRIMARY.ordinal();
+                    }
+                }));
+        }
+        /**
+         * (1) Reads are grouped by read group and read name.
+         * (2) Reads are then paired together as follows:
+         *   (a) The remaining reads (one per fragment key) are coordinate-sorted and paired
+         *       consecutively together and emitted.
+         *   (b) If a read is leftover, it is emitted, unmodified, as an unpaired end.
+         * (3) Paired ends are grouped by a similar key as above but using both reads.
+         * (4) Any unpaired end is emitted, unmodified.
+         * (5) The remained paired ends are scored and all but the highest scoring are marked as
+         *     duplicates. Both reads in the pair are emitted.
+         */
+        private PCollection<Read> transformReads(final PCollectionView<SAMFileHeader> headerPcolView, final PCollection<Read> pairs) {
+            final PCollection<KV<String, Iterable<Read>>> pairsGroupedByKey = groupPairsByKey(pairs, headerPcolView);
+            final PTransform<PCollection<? extends KV<String, Iterable<Read>>>, PCollection<KV<String, PairedEnds>>> markGroupedDuplicatePairs = markGroupedDuplicatePairs(headerPcolView);
+            final PTransform<PCollection<? extends KV<String, Iterable<PairedEnds>>>, PCollection<Read>> markPairedEnds = markPairedEnds();
+
+            return pairsGroupedByKey
+                    .apply(markGroupedDuplicatePairs)
+                    .setCoder(KvCoder.of(StringUtf8Coder.of(), PairedEndsCoder.of()))
+                    .apply(GroupByKey.<String, PairedEnds>create())
+                    .apply(markPairedEnds);
+        }
+
+        /**
+         * Groups pairs by keys - keys are tuples of (library, contig, position, orientation).
+         */
+        private PCollection<KV<String, Iterable<Read>>> groupPairsByKey(final PCollection<Read> pairs, final PCollectionView<SAMFileHeader> headerPcolView) {
+            final PTransform<PCollection<? extends Read>, PCollection<KV<String, Read>>> makeKeysForPairs =
+                    ParDo.named("make keys for pairs")
+                            .withSideInputs(Collections.singletonList(headerPcolView))
```

Comment:
you mean replace `.withSideInputs(Collections.singletonList(headerPcolView))` with `.withSideInputs(headerPcolView)`?
done